### PR TITLE
Rule: S4583 Calls to delegate's method "BeginInvoke" should be paired with calls to "EndInvoke"

### LINKED ISSUE #3965: Rule: S4583 Calls to delegate's method "BeginInvoke" should be paired with calls to "EndInvoke"
Implement https://jira.sonarsource.com/browse/RSPEC-4583

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #3965

### CODE REVIEW

File: analyzers/rspec/vbnet/S4583_vb.net.html [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+    Public Sub DoSomething()
+        Dim Example As New AsyncExample()
+        Dim Caller As New AsyncMethodCaller(Example.SomeMethod)
+        ' Initiate the asynchronous call.
+        Dim Result As IAsyncResult = Caller.BeginInvoke(New AsyncCallback(Sub(ar)
+                                                                              ' Call EndInvoke to retrieve the results.
+                                                                              Dim Ret As String = Caller.EndInvoke(ar)
+                                                                              ' ...
+                                                                          End Sub), Nothing) ' Noncompliant - Not paired With EndInvoke
```
Comment (csaba-sagi-sonarsource) [inline analyzers/rspec/vbnet/S4583_vb.net.html:73]: Why is this here in the Compliant section?

File: analyzers/src/SonarAnalyzer.CSharp/Helpers/CSharpMethodParameterLookup.cs [hunk 2]
```diff
@@ -23,22 +23,22 @@
...
-            : base(argumentList?.Arguments, argumentList == null ? null : semanticModel.GetSymbolInfo(argumentList.Parent).Symbol as IMethodSymbol)
-        {
-        }
+            : base(argumentList?.Arguments, argumentList == null ? null : semanticModel.GetSymbolInfo(argumentList.Parent).Symbol as IMethodSymbol) { }
 
         public CSharpMethodParameterLookup(ArgumentListSyntax argumentList, IMethodSymbol methodSymbol)
-            : base(argumentList?.Arguments, methodSymbol)
-        {
-        }
+            : base(argumentList?.Arguments, methodSymbol) { }
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Helpers/CSharpMethodParameterLookup.cs:38]: constructor can be single line

File: analyzers/src/SonarAnalyzer.Common/Facade/SyntaxFacade.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using Microsoft.CodeAnalysis;
+
+namespace SonarAnalyzer.Helpers
+{
+    public abstract class SyntaxFacade<TSyntaxKind>
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Facade/SyntaxFacade.cs:26]: Very nice idea, it will remove a lot of code from our codebase!

**Review by csaba-sagi-sonarsource — APPROVED** [review]
LGTM, I like the SyntaxFacade a lot!

File: analyzers/rspec/vbnet/S4583_vb.net.html (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/rspec/vbnet/S4583_vb.net.html:73]: Copy/Paste issue

File: analyzers/src/SonarAnalyzer.CSharp/Helpers/CSharpMethodParameterLookup.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Helpers/CSharpMethodParameterLookup.cs:38]: I intentionally put them on a separate line, as each of them contains specific logic. The longest one would produce too long line anyway and the file structure is regular like this.
